### DISCUSSION TITLE
Head Object Request on directory does not return object metadata.

### DISCUSSION THREAD

**Discussion by DynamicCommander** [minio/minio#13127] [Q&A]
## Expected Behavior

When calling stat in mc.exe on a directory the content type is folder and the size is 0B.
Name      : foo\
Size      : 0 B
Type      : folder
Metadata  :
  Versioning: Un-versioned
  Location: us-east-1
  Policy: none
  
  I expect a similar metadata return to happen when using Head Object Request in S3.

## Current Behavior

When calling a Head Object Request in S3 the error is returned RESOURCE_NOT_FOUND.

## Possible Solution

I expect a similar metadata return to happen when using Head Object Request in S3.

## Steps to Reproduce (for bugs)

1. Create a directory of directories and subdirectories 
    foo/
         --bar/
    folder/
         --folder2/
2. Set up a Head Object Request in AWS SDK S3 on directory foo/
3. Run Code
4. Head Object Outcome fails with error RESOURCE_NOT_FOUND

## Context

Trying to determine whether an item being requested from front-end is a singular file or an entire directory.

## Your Environment

* Version used (`minio --version`): minio.exe version RELEASE.2021-07-27T02-40-15Z
* Server setup and configuration: Standalone exe, using AWS SDK CPP
* Operating System and version (`uname -a`): Windows 10

**Comment by klauspost** [comment]
`mc` treats prefixes as folders, so it will try both headobject and listing.

This is by design. You can filter by `Type : folder`, if you want to exclude prefixes.

**Comment by DynamicCommander** [comment]
I'm sorry, I don't quite understand how your answer helps.

Aws::S3::Model::HeadObjectRequest request;
request.WithBucket("data").WithKey("foo/");

This returns RESOURCE_NOT_FOUND instead of 
Name : foo
Size : 0 B
Type : folder
Metadata :
Versioning: Un-versioned
Location: us-east-1
Policy: none

Like I think it should. 

Can you provide some more insight into what I may be doing wrong then?

**Comment by klauspost** [comment]
@DynamicCommander There exists no object named `foo/`, so `HeadObject` will always return `RESOURCE_NOT_FOUND`.

mc falls back to doing a listing (ListObjectsV2) with the `foo/` prefix. Since it gets a result it displays it as a folder.

**Comment by DynamicCommander** [comment]
If I do a ListObjectsV2 with prefix `foo/` 
I get the return value of `foo/bar/`

**Comment by klauspost** [comment]
@DynamicCommander That is how listing works.

https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html

**Comment by DynamicCommander** [comment]
I want to take the time to thank you for your help so far,

when I use a prefix of `foo/` with no delimiter I get `foo/bar` but,
if I use a prefix of `foo/` with a delimiter of `/` I get nothing. 

According to the documentation that you linked above:
sample.jpg
photos/2006/January/sample.jpg
photos/2006/February/sample2.jpg
photos/2006/February/sample3.jpg
photos/2006/February/sample4.jpg

Using this structure, with a prefix of `'photos/2006/'` and a delimiter of `/` should result in contents containing a key of `photos/2006/`

However, I am doing exactly the above implementation, and getting a result of 0 contents with a prefix of `photos/2006/` and a delimiter of `/` I am getting 0 contents where I should be getting the key of the directory itself. Correct?

And actually the example being used by AWS itself is more akin to the issue I'm having due to the folder structure being more similar with nested folders.

**Comment by harshavardhana** [comment]
> when I use a prefix of `foo/` with no delimiter I get `foo/bar` but,
> if I use a prefix of `foo/` with a delimiter of `/` I get nothing.

Can you share your code and also provide `mc tree` output?

**Comment by DynamicCommander** [comment]
![image](https://user-images.githubusercontent.com/10710698/131709829-fabe1232-9e1c-4c5f-aff9-c3c71a6d8a2e.png)

I can't share the full tree, but this directory is the one I'm working in. 

The output I'm expecting is a key of "BattleSpaces/BS Test 2" in the result of `ListObjectsV2(request)`

```
Aws::S3::Model::ListObjectsV2Request request;
request.WithBucket("data").WithPrefix("BattleSpaces/BS Test 2/").WithDelimiter("/");
auto outcome = s3_client->ListObjectsV2(request);
	Aws::Vector<Aws::S3::Model::Object> objects;

	if (outcome.IsSuccess())
		objects = outcome.GetResult().GetContents();
```

**Comment by jeremy-coulon** [comment]
Hello
Did you find a solution @DynamicCommander ?

I have the same need.

I think that minio is inconsistent in the way it supports directories.
Putting an empty object with a trailing slash "foo/" is supported by minio and returns success.
Then doing a HeadObject request on "foo/" fails with a 404 error.
I am expecting HeadObject to work on key with a trailing slash "/".

I did not try to use ListObjectsV2 yet. What is it supposed to return on an empty directory "foo/" ?
How do I know that an empty directory exists ?

**Reply by jeremy-coulon** [reply]
Do I have to request ListObjectsV2 on the parent directory that I am looking for and then iterate through results in order to check that the directory is present ?

For exemple:
* PutObject "foo/"
* PutObject "foo/bar/"
If I want to check that "foo/bar/" directory exists:
* ListObjectsV2 "foo/" should return "foo/bar/" (even is bar/ is an empty directory) ?
